cdap-api-spark/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContext.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.spark;
+
+import co.cask.cdap.api.RuntimeContext;
+import co.cask.cdap.api.ServiceDiscoverer;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.data.batch.Split;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.flow.flowlet.StreamEvent;
+import co.cask.cdap.api.metrics.Metrics;
+import co.cask.cdap.api.plugin.PluginContext;
+import co.cask.cdap.api.workflow.Workflow;
+import co.cask.cdap.api.workflow.WorkflowToken;
+import org.apache.spark.Partition;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.PairFunction;
+import scala.Tuple2;
+
+import java.io.Serializable;
+import java.nio.charset.Charset;
+import java.util.Collections;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Spark program execution context. User Spark program can interact with CDAP through this context.
+ */
+public abstract class JavaSparkExecutionContext implements RuntimeContext {
+
+  /**
+   * @return The specification used to configure this {@link Spark} job instance.
+   */
+  public abstract SparkSpecification getSpecification();
+
+  /**
+   * Returns the logical start time of this Spark job. Logical start time is the time when this Spark
+   * job is supposed to start if this job is started by the scheduler. Otherwise it would be the current time when the
+   * job runs.
+   *
+   * @return Time in milliseconds since epoch time (00:00:00 January 1, 1970 UTC).
+   */
+  public abstract long getLogicalStartTime();
+
+  /**
+   * Returns a {@link DatasetContext} to provide access to {@link Dataset}
+   * which can be passed in Spark program's closures.
+   *
+   * @return A {@link Serializable} {@link DatasetContext}
+   */
+  public abstract DatasetContext getDatasetContext();
```

Comment:
Because the returned object needs to be `Serializable` in order to be used inside Spark closure, while it is not the case in other contexts. If we make this context implements `DatasetContext`, we have to have the implementation implements Serializable, which is quite complicated to implement and to evolve. In fact, Spark users are also familiar with this model that the `SparkContext` itself is not serializable.